[CORE-8532]Datalake/Iceberg: Add support for primitive type promotion

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Expands support for schema evolution by introducing primitive type promotion.
Legal type promotions are as follows:

Primitive type | v1, v2 valid type promotions | v3+ valid type promotions | Requirements
-- | -- | -- | --
unknown |   | any type |  
int | long | long |  
date |   | timestamp, timestamp_ns | Promotion to timestamptz or timestamptz_ns is not allowed; values outside the promoted type's range must result in a runtime failure
float | double | double |  
decimal(P, S) | decimal(P', S) if P' > P | decimal(P', S) if P' > P | Widen precision only

Not addressed in this PR:
- Interactions with partition spec
- forward write compatibility 
   - producer updates are not synchronous w/ table metadata updates, so records published w/ a previous schema version will be dropped, though in many cases the write would be compatible with the new schema

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Iceberg/Datalake: Adds support for schema evolution by primitive type promotion

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
iceberg: Introduce compatibility module

Initially only supports checking for primitive type promotions.

iceberg::check_types(field_type s, field_type d) returns
  - type_promoted::no - if the types are identical (or non-primitive and equal)
  - type_promoted::yes - if s was promoted to d and the promotion is legal
  - an error - if s was promoted to d and the promotion is illegal
Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
datalake: Update catalog_schema_manager to support type promotion

Works similarly at the interface, but instead of checking for field-wise
type identity check for primitive type promotion. If one or more fields
was promoted, consider the resulting field assignment "incomplete" to
trigger a schema update operation.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59741-->
<!--8f88b300-d703-4130-ab92-929070753a66-->
<!--c43e4f0a-f640-4ea0-9448-38e7a58ac2a6-->
<!--title-->
## Retry command for [Build#59741](https://buildkite.com/redpanda/redpanda/builds/59741)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestCompactedTopic.test_compacting_during_leadership_transfer@{"cloud_storage_type":2}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--12195c6c-8b3e-4842-a5b8-f13be30c7a61-->
<!--TEST RESULTS-->
<!--8379dedc-bd18-4002-a934-2bbeeb7a45ac-->
<!--TEST RESULTS-->
<!--9a644cfb-adaf-4560-ab75-a1471b3574ba-->
<!--TEST RESULTS-->
<!--b01f22be-f65e-4eff-837f-d8dc3c2f2670-->
<!--TEST RESULTS-->
<!--72c56fa1-b126-460a-be3b-8460adcf36bd-->
<!--TEST RESULTS-->
<!--60e431a8-d046-4f68-a3bc-bb9374e27f33-->
<!--TEST RESULTS-->
<!--7970ee09-6491-4e78-b1cf-578a4491b181-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59741">build#59741</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=20.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59741#0193c20d-ac0e-4355-abfa-73166c5b0732 | FAIL | 0/6 |
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59741#0193c20d-ac0d-4ef0-a1de-679b47bbfdb9 | FAIL | 0/6 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTestCompactedTopic.test_compacting_during_leadership_transfer.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59741#0193c20d-ac0e-4355-abfa-73166c5b0732 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59741#0193c20d-ac0e-4355-abfa-73166c5b0732 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59741#0193c20d-ac0e-4355-abfa-73166c5b0732 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59741#0193c20d-ac0e-4355-abfa-73166c5b0732 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59741#0193c20d-ac0e-4355-abfa-73166c5b0732 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59786">build#59786</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59786#0193cd5e-d0be-49ae-9f3a-e4baf8ffa27a | FAIL | 0/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59829">build#59829</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59829#0193d1f8-a680-4209-9338-c4275cb61c1f | FLAKY | 4/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59921">build#59921</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=20.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59921#0193daf0-afcd-48dc-9af6-5d98f8c32d31 | FAIL | 0/6 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionTest.test_evolving_avro_schemas.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/59921#0193daeb-7e15-4df1-8905-28a44b0aad39 | FAIL | 0/1 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionTest.test_evolving_avro_schemas.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/59921#0193daf0-afcc-45b5-94ae-a6275b5f34ae | FAIL | 0/1 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionTest.test_evolving_avro_schemas.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO | ducktape | https://buildkite.com/redpanda/redpanda/builds/59921#0193daeb-7e16-4162-bafe-937b83565a28 | FAIL | 0/1 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionTest.test_evolving_avro_schemas.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO | ducktape | https://buildkite.com/redpanda/redpanda/builds/59921#0193daf0-afcc-4f13-8283-f7c95899208e | FAIL | 0/1 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTestCompactedTopic.test_write.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59921#0193daf0-afcd-48dc-9af6-5d98f8c32d31 | FLAKY | 2/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59930">build#59930</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=20.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59930#0193dbc0-ef3c-4c03-a415-3c50210c098e | FAIL | 0/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59930#0193dbc0-ef3c-4c03-a415-3c50210c098e | FLAKY | 4/6 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTestCompactedTopic.test_write.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59930#0193dbc0-ef3c-4c03-a415-3c50210c098e | FLAKY | 5/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59949">build#59949</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.prepared.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59949#0193de26-7351-407f-b621-696215d82c6a | FLAKY | 5/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59963">build#59963</a></summary>


| test_id                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=20.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59963#0193e01a-bbfc-449b-9f49-2f73735e1770 | FAIL          | 0/6      |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59963#0193e01a-bbfc-449b-9f49-2f73735e1770 | FLAKY         | 2/6      |
</details>
<!--/body-->

oleiman commented:
/ci-repeat 1
skip-units
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestCompactedTopic.test_compacting_during_leadership_transfer@{"cloud_storage_type":2}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}

oleiman commented:
/ci-repeat 1

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59786-->
<!--ba10b389-752a-41bb-a6aa-789c344c238b-->
<!--title-->
## Retry command for [Build#59786](https://buildkite.com/redpanda/redpanda/builds/59786)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

oleiman commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}

rockwotj commented on src/v/iceberg/compatibility.cc:
```diff
@@ -0,0 +1,104 @@
+        return compat_errc::mismatch;
+    }
+
+    template<typename T>
+    type_check_result operator()(const T&, const T&) const {
+        return type_promoted::no;
+    }
```
Worth a comment that changes to struct field member set is handled elsewhere?

oleiman replied:
there's a comment in the unit test, but that's a silly place for it. will move.

oleiman replied:
done

rockwotj commented on src/v/datalake/catalog_schema_manager.cc:
```diff
@@ -38,8 +40,11 @@ enum class fill_errc {
     // There is a mismatch in a field's type, name, or required.
     mismatch,
     // We couldn't fill all the columns, but the ones we could all matched.
+    // Or one or more columns were type-promoted (legally), so we'll need to
+    // push an update into the catalog.
```
Can we rename the options now since these don't make sense to me? something like `invalid_schema` and `schema_evolution_needed`?

oleiman replied:
yeah, i meant to put a note somewhere - my thought was that this whole private API will basically evaporate when I add struct evolution, so the code motion feels like noise. i agree it's unsightly tho; will just change it.

oleiman replied:
done

rockwotj reviewed (commented):
Generally looks awesome thanks! I've just got a stocking of nits for you.

rockwotj commented on src/v/iceberg/compatibility.h:
```diff
@@ -0,0 +1,29 @@
+
+namespace iceberg {
+
+enum class compat_errc {
+    mismatch,
+};
+
+using type_promoted = ss::bool_class<struct type_promoted_tag>;
+using type_check_result = checked<type_promoted, compat_errc>;
+
+type_check_result
+check_types(const iceberg::field_type& src, const iceberg::field_type& dest);
```
nit: some light docs would be 😎

oleiman replied:
yeah, will add in a followup. larger PR for add/remove/reorder not far behind

andrwng replied:
+1, it'd be great to call out here that this is primarily checking the primitives and isn't expected to be used as a standalone nested type-promotion checker

oleiman replied:
done

rockwotj reviewed (dismissed):
LGTM - thanks! I really like the usage of template overloads for the compatible type checks.

andrwng commented on src/v/datalake/catalog_schema_manager.cc:
```diff
@@ -56,23 +61,28 @@ fill_field_ids(iceberg::struct_type& dest, const iceberg::struct_type& source) {
+            || compatibility.has_error()) {
+            return fill_errc::invalid_schema;
+        } else if (compatibility.value() == type_promoted::yes) {
+            has_primitive_type_promotion = true;
```
Could we have just returned early here? Or is the intent that we still fill the field IDs when there is a promotion?

oleiman replied:
At this stage, I think you still want to know whether filling the IDs completes, as before. The intent is to widen the type equality requirement w/o changing any other assumptions

andrwng replied:
Ah 🤦 yea makes sense. Thanks!

andrwng reviewed (dismissed):
LGTM! Wondering if the contents of this PR are already something that can be validated with a ducktape test, or whether there's additional changes needed to work end-to-end?

oleiman commented:
> LGTM! Wondering if the contents of this PR are already something that can be validated with a ducktape test, or whether there's additional changes needed to work end-to-end?

Should work e2e. I was under the impression that DT support was still largely WIP. If it's already possible to write integration tests, I'll do that today.

andrwng commented:
> > LGTM! Wondering if the contents of this PR are already something that can be validated with a ducktape test, or whether there's additional changes needed to work end-to-end?
> 
> Should work e2e. I was under the impression that DT support was still largely WIP. If it's already possible to write integration tests, I'll do that today.

Yep ducktape tests exist today with different catalogs! https://github.com/redpanda-data/redpanda/blob/11cda5c9c63043fa89d7eb279baf24cbee57b365/tests/rptest/tests/datalake/datalake_e2e_test.py#L96 is an example of one that uses a schema, though for evolution maybe it'd be easier to use the Kafka python API directly

oleiman committed:
dt/datalake: Basic integration test for schema evolution

Uses RPCN + filesystem catalog to check that a simple primitive type
promotion behaves as expected, end-to-end.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/9fc1b84a19fa862f0f785af10098fb431374ae09..764c00ca599ab7586fd1c19a2ed6e8393b698b5a) ducktape and docstring

oleiman commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -0,0 +1,207 @@
+            tf.flush()
+            rpk.create_schema(subject, tf.name)
+
+    @cluster(num_nodes=4, log_allow_list=["Schema \d+ already exists"])
```
> `log_allow_list...`

@andrwng - this occurs on maybe 5% of runs. workers racing on the in-memory metadata or something like that? seems benign with respect to this PR, but flagging here in case it's surprising to you.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59921-->
<!--94349e1d-f42a-4d55-85d3-3379b7c6d9c5-->
<!--31784e87-fc32-42a2-82ba-ca103efc57d1-->
<!--8ddf7be7-a72f-4f5f-8f96-6def99c89ee5-->
<!--66bf172f-3f21-4049-80db-0d75b0f6df79-->
<!--23a116ae-37da-4e72-a1a7-9db278343877-->
<!--title-->
## Retry command for [Build#59921](https://buildkite.com/redpanda/redpanda/builds/59921)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionTest.test_evolving_avro_schemas@{"cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionTest.test_evolving_avro_schemas@{"cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}
```
<!--/body-->

oleiman commented:
CI Failure:
- https://github.com/redpanda-data/redpanda/pull/24561#issuecomment-2552075894 looks like my local is out of sync w/ dev 😎

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/764c00ca599ab7586fd1c19a2ed6e8393b698b5a..811af49e156494af8a937779f9a0590f21c59ae8) rebase dev and update validator assert message in `expect_exception`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59930-->
<!--9a86c7a3-486c-49fa-aedc-e16817ef7939-->
<!--title-->
## Retry command for [Build#59930](https://buildkite.com/redpanda/redpanda/builds/59930)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}
```
<!--/body-->

oleiman commented:
CI Failure:
- https://github.com/redpanda-data/redpanda/pull/24561#issuecomment-2552506135 - unrelated cloud retention flake

oleiman commented:
/ci-repeat 1
skip-units
skip-redpanda-build
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}

rockwotj commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -0,0 +1,206 @@
+                "processors": [{
+                    "mapping":
+                    f"""
+                            root.verifier_string = {mapping.get('verifier_string', 'uuid_v4()')}
```
Why not write this directly in the generate input?

oleiman replied:
Idk...copy pasta from another similar test. Maybe I should learn how to use connect 😅

oleiman replied:
done

rockwotj commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -0,0 +1,206 @@
+    "name": "VerifierRecord",
+    "fields": [
+        {
+            "name": "verifier_string",
```
Would be cool to test the existing schema evolution andrew added for example, adding a column and marking a required column null (because it's gone).

oleiman replied:
I feel like that stuff is covered pretty well by fixture tests. I'm treating this more as smoke test, with an explicit plan to build out a more detailed integration suite once the struct evolution PRs go in and we iron out product details. Sound ok to you?

rockwotj replied:
Yeah I guess that's fine. I am asking for smoke tests that work for adding and removing columns but as long as they are added at some point I am fine to wait.

oleiman replied:
included add/remove column cases. in latest diffs

rockwotj reviewed (commented):
Looks awesome! A stocking of nits for you. Ho ho ho 🎅

Another thought I had, we support widening types, but if the user narrows them again (or in the first place) should we widen the type again? Example:

User has an int32 proto type, they upgrade it to int64, then a new schema uses int32 again. We should upcast the int32 to int64 in the new schema right? Because the historical iceberg data that was int32 was upcast to int64 - should we do that for new data? 🤔

oleiman commented:
> Looks awesome! A stocking of nits for you. Ho ho ho 🎅
> 
> Another thought I had, we support widening types, but if the user narrows them again (or in the first place) should we widen the type again? Example:
> 
> User has an int32 proto type, they upgrade it to int64, then a new schema uses int32 again. We should upcast the int32 to int64 in the new schema right? Because the historical iceberg data that was int32 was upcast to int64 - should we do that for new data? 🤔

Yeah I've been thinking about this a lot. Similar reasoning applies to a write from an out-of-sync producer I think - in principle, and depending on how we constrain evolution, old-schema records could be write-compatible with a new schema table, right? Maybe that's a stretch.

oleiman commented:
/ci-repeat 1
skip-units
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/811af49e156494af8a937779f9a0590f21c59ae8..820db9a5a73673e57143c2896973baf949075c4a) bit better connect usage

rockwotj commented:
> Yeah I've been thinking about this a lot. Similar reasoning applies to a write from an out-of-sync producer I think - in principle, and depending on how we constrain evolution, old-schema records could be write-compatible with a new schema table, right? Maybe that's a stretch

Absolutely. I was thinking about the best way to model this. If we have already started writing a parquet file with the old schema we can add to that, but if we have started writing a parquet file with the new schema, then we likely want to upcast the data into the new schema just like the query engine will do for historical data. I think that means we need an extra "is this schema compatible" check, then a bunch of transforms to apply to a given schema to transform data into the latest table schema.

rockwotj commented:
Anyways that can be another PR. This LGTM

rockwotj commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -0,0 +1,216 @@
+        cloud_storage_type=supported_storage_types(),
+        query_engine=[
+            QueryEngineType.SPARK,
+            # QueryEngineType.TRINO,
```
?

oleiman replied:
sigh. added a remove column test case to dull the pain.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/820db9a5a73673e57143c2896973baf949075c4a..672e64a32981d831f07f7dfffd0f3a40db19ce18) - re-enable trino test variant and add a case for dropping a column

oleiman commented:
>upcast the data into the new schema just like the query engine will do for historical data.

Yeah exactly

> I think that means we need an extra "is this schema compatible" check, then a bunch of transforms to apply to a given schema to transform data into the latest table schema.

That's what I've been working on for the last couple days, in the forward direction. Basically we recursively visit the two schemas in lockstep, generate a "recipe" of transformations, then evaluate the recipe to produce a compatible schema. It should work in both directions, though some transformations might not be bidirectional. Fallback in either case can be DLQ (or whatever).

Applies for reordered and added/removed columns as well. e.g. 
- if a new column is optional, the upcast writes a null there. if it has a write default, the upcast writes that.
- if the order of columns changed, map the column ID by name. 
- when we "remove" a column, keep it around but make it nullable.

rockwotj commented:
:chefs_kiss:

Sounds perfect. Looking forward to it.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59963-->
<!--97295007-0d89-4d1b-b21e-ed8f871fa20d-->
<!--title-->
## Retry command for [Build#59963](https://buildkite.com/redpanda/redpanda/builds/59963)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}
```
<!--/body-->

oleiman commented:
/ci-repeat 1
skip-units
skip-redpanda-build
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}

oleiman commented:
/ci-repeat 1
skip-units
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}

rockwotj commented:
That test is flaky rn


Happy to force merge if that is the only one and you deflaked your test in CI

oleiman commented:
> That test is flaky rn
> 
> Happy to force merge if that is the only one and you deflaked your test in CI

Ok, cool. Still need to fire off a dt-repeat run on the new test, so fine to hold off merging.

oleiman commented:
/ci-repeat 3
skip-units
tests/rptest/tests/datalake/schema_evolution_test.py
dt-repeat=50

oleiman commented:
/ci-repeat 3
skip-units
tests/rptest/tests/datalake/schema_evolution_test.py
dt-repeat=50

oleiman commented:
[ducktape-release](https://buildkite.com/redpanda/redpanda/builds/59978#0193e1e0-6488-4adb-9d0b-2eb1ab566ece) stalled out at infra init, tests never ran

oleiman commented:
@rockwotj - I think we're good here if you don't mind force merging. Failed ducktape-release state is misleading & no flake afaict
